feat: restructure and clean up streaming async sampling (#1130)

Freder-chen opened the pull request.

hijkzzz commented:
TODO: I have an idea to simplify the code: for the AsyncLLM or agent scenario, we can just support streaming sampling. This way, we can remove a lot of code, and I think the code can be further simplified.

Freder-chen commented:
**Pretty burned out after this one, but everything is finally in place. 🔥**

- Rebuilt sampling around `RemoteSampleGenerator` and Ray `RolloutWorker` to drive vLLM/agent/remote-RM rollouts, with dynamic filtering, pass-rate reporting, and vLLM sleep/wake handling.
- Refactored PPO trainers (sync + async) onto the new sampler: cleaner logging (wandb/TensorBoard), resumable dataloader state, dynamic batch balancing, and async buffer/signal flow control so streaming generation and weight updates don’t collide.
- Simplified experience making into a single path: unified reward/advantage computation (reward model or precomputed rewards), overlong-response penalties, group reward stats, and normalized advantages.
- Consolidated vLLM actors into one async engine (logprobs mode, collective sync) plus rollout worker creation helpers; expanded CLI/config for async queue sizing, remote RM/agent hooks, and dynamic filtering, with a new streaming training script and refreshed agent examples.

hijkzzz commented on openrlhf/trainer/ray/rollout_worker.py:
```diff
@@ -0,0 +1,299 @@
+        elif remote_rm_url:
+            self.executor = RewardedGenerationExecutor(remote_rm_url, remote_rm_batch_size)
+        else:
+            self.executor = GenerationExecutor()
```
RewardModel

hijkzzz commented on openrlhf/trainer/ppo_trainer_async.py:
```diff
@@ -1,28 +1,59 @@
 
 
+@ray.remote(num_cpus=0)
+class BufferActor:
```
不要封装

Freder-chen commented on openrlhf/trainer/ppo_trainer_async.py:
```diff
@@ -249,104 +206,79 @@ def __init__(
-            remote_reward_model=self.remote_reward_model,
-            **generate_kwargs,
+            rollout_queue=self.rollout_queue,
+            rollout_slots=self.rollout_slots,
```
delete

Freder-chen commented on openrlhf/trainer/ppo_utils/experience_maker.py:
```diff
@@ -205,251 +210,297 @@ def concat_experiences(experiences_list: List["Experience"], pad_token_id) -> "E
+        if exhausted:
+            return [], prompts_consumed, exhausted
+
+        pending_refs = self._dispatch_prompts(prompts, labels, **generate_kwargs)
```
_dispatch_prompt_to_vllm

Freder-chen commented on openrlhf/trainer/ray/vllm_engine.py:
```diff
@@ -1,112 +1,177 @@
 
 
 @ray.remote
 class LLMRayActor(BaseLLMRayActor):
```
merge base actor
